examples/spells/BoundedUInt.qnt

```diff
@@ -0,0 +1,607 @@
+  pure def checkedSub(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, checkedSubUnsafe)
+
+  /// Unsafe checked integer multiplication. 
+  /// Computes `lhs * rhs`, setting the error field to "overflow" if overflow occurred.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def checkedMulUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v * rhs.v
+    {
+      v: res,
+      error: if (res > MAX) "overflow" else ""
+    }
+  }
+
+  /// Checked integer multiplication. 
+  /// Computes `lhs * rhs`, setting the error field to "overflow" if overflow occurred.
+  pure def checkedMul(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, checkedMulUnsafe)
+
+  /// Unsafe checked integer division. 
+  /// Computes `lhs / rhs`. If `rhs.v == 0` the error field is set to "division by zero" and the value is arbitrary.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def checkedDivUnsafe(lhs: T, rhs: T): T = 
+    if (rhs.v == 0) { v: 0,             error: "division by zero" }
+    else            { v: lhs.v / rhs.v, error: ""                 }
+  
+  /// Checked integer division. 
+  /// Computes `lhs / rhs`, setting the error field to "division by zero" if `rhs.v == 0`.
+  pure def checkedDiv(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, checkedDivUnsafe)
+
+  /// Unsafe checked integer remainder. 
+  /// Computes `lhs % rhs`. If `rhs.v == 0` the error field is set to "division by zero" and the value is arbitrary.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def checkedRemUnsafe(lhs: T, rhs: T): T = 
+    if (rhs.v == 0) { v: 0,             error: "division by zero" }
+    else            { v: lhs.v % rhs.v, error: ""                 }
+
+  /// Checked integer remainder. 
+  /// Computes `lhs % rhs`. If `rhs.v == 0` the error field is set to "division by zero" and the value is arbitrary.
+  pure def checkedRem(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, checkedRemUnsafe)
+
+  /// Unsafe checked exponentiation. 
+  /// Computes `lhs ^ rhs`, setting the error field to "overflow" if overflow occurred.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def checkedPowUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v ^ rhs.v
+    {
+      v: res,
+      error: if (res > MAX) "overflow" else ""
+    }
+  }
+
+  /// Checked exponentiation. 
+  /// Computes `lhs ^ rhs`, setting the error field to "overflow" if overflow occurred.
+  pure def checkedPow(lhs: T, rhs: T): T = wrapErrorBin(lhs, rhs, checkedPowUnsafe)
+
+    ///////////////////////////
+   // SATURATING OPERATIONS //
+  ///////////////////////////
+
+  /// Unsafe saturating integer addition. 
+  /// Computes `lhs + rhs`, saturating at the numeric bounds instead of overflowing.
+  /// Assumes lhs and rhs are valid (w.r.t. `isValid`).
+  pure def saturatingAddUnsafe(lhs: T, rhs: T): T = {
+    val res = lhs.v + rhs.v
+    { 
+      v: if (res < MAX) res else MAX, 
+      error: "" 
+    }
+  }
```

Comment:
Since these don't error out, would it not make more sense to just have it defined over `int`?

Reply:
I thought the same thing. Seems to me that we'd want all `saturating` and `wrapping` operators to be defined over `int`.

Reply:
It's an interface question. You want these operations to all be `(UIntT, UIntT) => UIntT` so you can chain them. 
If this returned an int, you couldn't do e.g. `saturatingAdd(saturatingAdd(a,b), c)`

Reply:
We are thinking the signature should be `(int, int) => int`, so that would chain fine.

Reply:
For `saturating` alone, yes, but you couldn't combine e.g. `saturatingAdd(checkedAdd(a,b), c)` (which was supposed to be the example above, but I borked it in copypaste)

Reply:
That makes sense, but is it something people would use? I can't really imagine a scenario where someone would perform a checked operation and then wrapping/saturation on top of that.

Reply:
The underlying issue is, that Rust uses e.g. `u64` and `Option<u64>` in these ops:
```rust
pub const fn checked_add(self, rhs: u64) -> Option<u64>
pub const fn saturating_add(self, rhs: u64) -> u64
```
which don't quite translate 1-to-1 to our `int` and `T = {v: int, error: str}` types. Specifically, operations which take `u64` as arguments never have to worry about _invalid_ inputs. In this sense, `T` is more than just an option type, since it also plays the role of enforcing (or at least having an interface which allows for the checking of) boundary constraints.
This is not an issue for the wrapping and saturating operations,  _however_, it does mean that `checked` needs to have its arguments either be `T`-typed, or perform input validation on `int` values 
(see the thread under https://github.com/informalsystems/quint/pull/1230#discussion_r1376820888 for examples).

In the rust interface, `x.checkedAdd(y.saturatingAdd(z))` is possible, so if we want it to also be possible here, we need to either have:
```
saturatingAdd: (T,T) => T
checkedAdd: (T,T) => T
```
where `T` plays the role of a hybrid option-and-no-exception-thrown type, or
```
saturatingAdd: (int, int) => int
checkedAdd: (int, int) => T
```
where `T` is purely an option type, but we have to add `int`-validation to every method, since `saturatingAdd(-1,-1)` could never be called in rust, as `-1` is not a `u64` value.

Reply:
As I understand it, you two seem to prefer the 
```
saturatingAdd: (int, int) => int
checkedAdd: (int, int) => T
```
scenario, so I'll go ahead and change it to behave that way

Reply:
Although, now that I think about it, I'm not sure I know how to combine input validation and an `int` return type. In `saturatingAdd(-1,-1)` I don't see how you'd alert the user that `-1` is an invalid input. Nor does it make the `intToUInt` constructor clear or sensible.

Reply:
If the aim of this module is to follow the Rust lib as closely as possible, and they also the same underlying type for saturating and checked values, that's a good reason to keep it, IMO. But please document that motivation, including a link to a motivating rust docs or code.

> This is not an issue for the wrapping and saturating operations, however, it does mean that checked needs to have its arguments either be T-typed, or perform input validation on int values

IMO, this should maybe motivate us to consider introducing type abstraction. This would allow us to prevent construction of values of type `T` by means other than the constructors provided in this module and would solve the needless validation problem.

Reply:
Hmm, lets maybe wait for @konnov and @thpani  to get back and weigh in on this.

Reply:
I prefer easy operator chaining over compatibility with the Rust library.
Our goal here should be brief, self-explanatory specs.

Reply:
yeah, chaining would be good IMO

Reply:
Just to make sure I understand the above, you also prefer
```
saturatingAdd: (int, int) => int
checkedAdd: (int, int) => T
```
?
We can do it like that, but I'll just reiterate https://github.com/informalsystems/quint/pull/1230#discussion_r1380019675, doing this allows for "nonsensical" (in the unsigned-int universe) constructions.

Reply:
Both ways are "chainable"

Reply:
> Our goal here should be brief, self-explanatory specs

I agree with that goal! But its worth noting that requiring the `saturatinAdd` arguments and return value to be wrapped introduces its own complexity, and leaving them as type `int` makes it clearer when interacting with other int operators, like say the predicates or indexing operations.

Wrapping everything in the `UInt` type, whether or not it is needed, will make it easier to chain these operators and that is clearer, but at the expense of every other point where the returned or supplied values have to be cast to/from an int.

An additional benefit of expecting this to use `Option` combinators for chaining (once sum types are in, but using a makeshift maybe in the meantime) is that it makes it self-explanatory to readers that one is chaining partial computations.

Some example use cases using the two different approaches may be illustrative to track the tradeoffs. Do we have any code in mind that would be a good test case?

Reply:
At least from the code where this would be useful that I've seen (Rust in Cosmos, bounded integers in smart contracts), you would firmly stay within the bounded fragment. I've never seen an unwrap to integers.

For this purpose, I believe that a single, initial wrapping of Quint's native `int` into a `UInt`, and then simply remaining in `UInt`, is the most practical.

But I'm happy to consider other examples, or if we have none atm, revise this library when necessary.

Reply:
Makes sense.